pallets/ethereum-transaction/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cfg_traits::ethereum::EthereumTransactor;
+use ethereum::{LegacyTransaction, TransactionAction, TransactionSignature, TransactionV2};
+use fp_evm::CallOrCreateInfo;
+use frame_support::{
+	dispatch::{DispatchErrorWithPostInfo, PostDispatchInfo},
+	pallet_prelude::*,
+};
+pub use pallet::*;
+use pallet_evm::{ExitError, ExitFatal, ExitReason};
+use sp_core::{H160, H256, U256};
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+const TRANSACTION_RECOVERY_ID: u64 = 42;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_ethereum::Config {
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	/// Storage for nonce.
+	#[pallet::storage]
+	pub(crate) type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A call was executed.
+		Executed {
+			from: H160,
+			to: H160,
+			exit_reason: ExitReason,
+			value: Vec<u8>,
+		},
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn get_transaction_signature() -> Option<TransactionSignature> {
+			//TODO(cdamian): Same signature as the one in ethereum-xcm.
+			TransactionSignature::new(
+				TRANSACTION_RECOVERY_ID,
+				H256::from_low_u64_be(1u64),
+				H256::from_low_u64_be(1u64),
+			)
+		}
+	}
+
+	impl<T: Config> EthereumTransactor for Pallet<T> {
+		/// This implementation serves as a wrapper around the Ethereum pallet
+		/// execute functionality. It keeps track of the nonce used for each
+		/// call and builds a fake signature for executing the provided call.
+		///
+		/// NOTE - The execution fees are charged by the Ethereum pallet,
+		/// we only have to charge for the nonce read operation.
+		fn call(
+			from: H160,
+			to: H160,
+			data: &[u8],
+			value: U256,
+			gas_price: U256,
+			gas_limit: U256,
+		) -> DispatchResultWithPostInfo {
```

Comment:
NOTE - this was built with the idea that the caller of this function would be the one that will take of charging the according fees based on the returned weight. This was an ongoing discussion a few weeks back and I think we should get back to it and decide how we want to tackle this.

Reply:
Agreed. We can add an additional parameter if we want to let the system pay I think!